### DISCUSSION TITLE
Add/Update GitHub Actions

### DISCUSSION THREAD

**Discussion by daPhipz** [Schneegans/Fly-Pie#107] [Ideas & Feature Discussions]
For some time now, I contribute German translations to the [GNOME project](https://l10n.gnome.org/languages/de/). Recently, one of my translations failed tests when uploading  - I think `msgfmt -vc` was the command they run when testing the uploaded translations (can't confirm it, I'm not on my coding machine(tm) for the next week). I think this would be cool for us as well!

While I'm at it - do you have ideas for other CI tests we should do? Do any existing tests need updating?
I will start working on it next weekend, so plenty of time to think ;)

**Comment by Schneegans** [comment]
I thought about this - any meaningful code-related tests would involve launching GNOME Shell on a GitHub hosted runner. While this should be close-to-impossible, I already came quite far!

I managed to created a Podman container (https://github.com/Schneegans/gnome-shell-pod) which executes `gnome-shell` under `xvfb`. It's possible to run this container on a GitHub runner and upload a screenshot of GNOME Shell as artifact! See one of the latest CI runs: https://github.com/Schneegans/Fly-Pie/actions/runs/514312628. If you scroll to the bottom of the page, you can download the screenshot.

The only missing puzzle piece is that GNOME Shell fails to load any extension when run this way. I asked for any ideas regarding this issue over at https://discourse.gnome.org/t/run-gnome-shell-in-a-container/5450. Let's see if the folks have an idea!

**Comment by Schneegans** [comment]
Extension loading seems to work now! Feel free to test it on your machine to get a feeling for the possibilities!

https://github.com/Schneegans/gnome-shell-pod

I think we could do _alot_ of testing this way. We can actually simulate mouse clicks, open the settings dialog etc...

**Comment by daPhipz** [comment]
Nice! Will test it on the weekend if there's time (I am working this weekend), or in the coming week.

**Comment by Schneegans** [comment]
It's finally working! We now have a test which executes this: https://github.com/Schneegans/Fly-Pie/blob/develop/scripts/run-tests.sh.

Feel free to test this on your end, you'll only need podman (`sudo apt install podman`). The script installs Fly-Pie in GNOME Shell 3.36 or 3.38 (or 40.0 but that's not yet supported in `develop`) and opens the preferences dialog. If any error happens, the script reports a failure.

Now we can extend this to whatever we want:
* Import menu configurations from older versions.
* Actually opening menus.
* Opening menus via the D-Bus.
* Simulating mouse clicks in the preferences dialog.
* ...

You may have a look at https://github.com/Schneegans/gnome-shell-pod to get some inspiration! I'll make this a discussion; we should then later create issues for specific tests.
